File: src/2025-02-23-deepclaude-try/java/exercises/practice/affine-cipher/src/main/java/AffineCipher.java
public class AffineCipher {
    
    public String encode(String text, int coefficient1, int coefficient2) {
        validateCoefficient1(coefficient1);
        String processed = processEncode(text, coefficient1, coefficient2);
        StringBuilder grouped = new StringBuilder();
        int len = processed.length();
        int groupPos = 0;
        for (int i = 0; i < len; i++) {
            if (groupPos == 5) {
                grouped.append(' ');
                groupPos = 0;
            }
            grouped.append(processed.charAt(i));
            groupPos++;
        }
        return grouped.toString();
    }

    public String decode(String text, int coefficient1, int coefficient2) {
        validateCoefficient1(coefficient1);
        int mmi = modInverse(coefficient1);
        String processedText = text.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        return processDecode(processedText, coefficient1, coefficient2, mmi);
    }

    private static void validateCoefficient1(int a) {
        if (gcd(a, 26) != 1) {
            throw new IllegalArgumentException("Coefficient1 and 26 must be coprime.");
        }
    }

    private static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private static int modInverse(int a) {
        a = a % 26;
        for (int x = 1; x < 26; x++) {
            if ((a * x) % 26 == 1) {
                return x;
            }
        }
        throw new IllegalArgumentException("No modular inverse exists.");
    }

    private static String processEncode(String text, int a, int b) {
        StringBuilder result = new StringBuilder();
        for (char ch : text.toCharArray()) {
            if (Character.isLetter(ch)) {
                char lowerCh = Character.toLowerCase(ch);
                int x = lowerCh - 'a';
                int encrypted = (a * x + b) % 26;
                if (encrypted < 0) {
                    encrypted += 26;
                }
                result.append((char) (encrypted + 'a'));
            } else if (Character.isDigit(ch)) {
                result.append(ch);
            }
        }
        return result.toString();
    }

    private static String processDecode(String text, int a, int b, int mmi) {
        StringBuilder result = new StringBuilder();
        for (char ch : text.toCharArray()) {
            if (Character.isLetter(ch)) {
                char lowerCh = Character.toLowerCase(ch);
                int y = lowerCh - 'a';
                int decrypted = (mmi * (y - b)) % 26;
                if (decrypted < 0) {
                    decrypted += 26;
                }
                result.append((char) (decrypted + 'a'));
            } else if (Character.isDigit(ch)) {
                result.append(ch);
            }
        }
        return result.toString();
    }
}

File: src/2025-02-23-deepclaude-try/python/exercises/practice/affine-cipher/affine_cipher.py
import math

def _is_coprime(a):
    return math.gcd(a, 26) == 1

def _mod_inverse(a):
    for x in range(1, 26):
        if (a * x) % 26 == 1:
            return x
    raise ValueError("No inverse exists for a mod 26")

def encode(plain_text, a, b):
    if not _is_coprime(a):
        raise ValueError("a and m must be coprime.")
    processed = [c.lower() for c in plain_text if c.isalnum()]
    ciphered = []
    for c in processed:
        if c.isdigit():
            ciphered.append(c)
        else:
            x = ord(c) - ord('a')
            encrypted_x = (a * x + b) % 26
            encrypted_c = chr(encrypted_x + ord('a'))
            ciphered.append(encrypted_c)
    ciphered_str = ''.join(ciphered)
    grouped = ' '.join([ciphered_str[i:i+5] for i in range(0, len(ciphered_str), 5)])
    return grouped

def decode(ciphered_text, a, b):
    if not _is_coprime(a):
        raise ValueError("a and m must be coprime.")
    a_inv = _mod_inverse(a)
    processed = [c.lower() for c in ciphered_text if c.isalnum()]
    decrypted = []
    for c in processed:
        if c.isdigit():
            decrypted.append(c)
        else:
            y = ord(c) - ord('a')
            decrypted_x = (a_inv * (y - b)) % 26
            decrypted_c = chr(decrypted_x + ord('a'))
            decrypted.append(decrypted_c)
    return ''.join(decrypted)
